### PR TITLE
stop client websocket close echoing server close

### PR DESCRIPTION
**Pull request by bobthekingofegypt**
Client initiated close requests should send a close message to the
server that the server will echo back to complete the process.  Servo
should not then echo the servers close request back again to the server,
this guard stops servo from echoing a server close request if the
process was initiated by the client.

Tracked in https://github.com/servo/servo/issues/9803#issuecomment-196424406

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10382)

### CODE REVIEW

File: components/net/websocket_loader.rs [hunk 1]
```diff
@@ -123,7 +126,9 @@ pub fn init(connect: WebSocketCommunicate, connect_data: WebSocketConnectData, c
                     },
                     Type::Pong => continue,
                     Type::Close => {
-                        ws_sender_incoming.lock().unwrap().send_message(&message).unwrap();
+                        if client_initiated_close_incoming.load(Ordering::SeqCst) == false {
```
Comment (KiChjang) [inline components/net/websocket_loader.rs]: nit: `if !client_initiated_close_incoming.load(Ordering::SeqCst) {`

**Comment by metajack** [comment]
r=me with @kichjang's comment addressed

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10382#-:-KEX685h4XBQXo7iZ2ED:-904289873)*

**Comment by metajack** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10382)*

**Comment by bobthekingofegypt** [comment]
@KiChjang's comment addressed.

**Comment by metajack** [comment]
@bors-servo r+

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10382#-:-KEXaTWK4MTTDNZy5N9W:1236647940)*

**Comment by metajack** [comment]
@bors-servo r-

Please squash and we'll get this merged.

+S-needs-squash 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10382#-:-KEXaevTZ_Foi7yeqFVy:351688215)*

**Comment by bobthekingofegypt** [comment]
Sorry, I'll need to add the guard around https://github.com/bobthekingofegypt/servo/blob/fix-websocket-close/components/net/websocket_loader.rs#L160 as well.  As it's still possible to send two close messages.  Will get it done tomorrow.

**Comment by bobthekingofegypt** [comment]
Sorry for the delay, I added the check around both incoming and outgoing.  If it's ok, I'll squash all the comimits.

**Comment by metajack** [comment]
Looks good, aside from the comment I left. You can go ahead and squash it at the same time you address that.

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/net/websocket_loader.rs, line 129 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10382#-KFG3RD8l7-4VWGITuTT:-KFG1e6iAxGvRyYSx4ws:-1788352352) ([raw file](https://github.com/servo/servo/blob/97c58fc6d097eca85ca097036e9a83fbbf63508c/components/net/websocket_loader.rs#L129)):*
If we weren't waiting for the close to come in, we now set that we are. When we aren't waiting for a close, getting a close needs a ack. By sending this ack, we should not be getting an ack to our ack, nor did we send a client intiated close. So I'm not sure why you are toggling the bool in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10382#-:-KFG32NAk3eJ5q6BrL-V:-1174653106)*

**Comment by perlun** [comment]
@bobthekingofegypt - any progress on this one?

**Comment by bobthekingofegypt** [comment]
I can't really use computers at the moment due to an issue with my eye.  So I can't really deal with this right now.  I hope to be back to normal in a few weeks.

In response to your comment @metajack the reason that the boolean is checked and updated on both incoming and outgoing threads is due to the current threading.
From memory without checking anything it is to do with the websocket server initiating a close and the user script on servo side calling close at the same time, happens a lot on the test cases.  
Whatever happens you only want to send one close whether it is an initiator or an echo, the boolean check and update on both means only one close with ever be sent regardless of the order the threads are executed.

I can write up a better explanation when I'm back but I'm just rushing this comment for now.  If someone else wants to wrap this up feel free.

**Comment by bobthekingofegypt** [comment]
*[components/net/websocket_loader.rs, line 129 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10382#-KFG3RD8l7-4VWGITuTT:-KH68hPkk9pMOaWd_7fN:792466469) ([raw file](https://github.com/servo/servo/blob/97c58fc6d097eca85ca097036e9a83fbbf63508c/components/net/websocket_loader.rs#L129)):*
Sorry, I think I misunderstood this when I replied in the PR.  I'm still not sure I fully understand.  I think I may have confused things by naming the variable initiated_close rather than something like sent_close.
I flip the guard here as well as down below because I only want to send a close once, regardless of whether it is an echo or an instigator.  If I don't flip the guard here then a JS call to close the websocket can still trigger another close event through line 155 as the state on dom/websocket side is not yet marked as closed or closing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10382)*

**Comment by jdm** [comment]
The code and explanation makes sense to me. @bobthekingofegypt if you squash the commits, I'll merge it :)

**Comment by metajack** [comment]
I'm not sure how I forgot to comment here. I could have sworn I wrote out an r=me.

In any case, I apologize for the delay and I agree with jdm.

**Pull request merged**